/*
 * Copyright 2019 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";

package updatable_private_set_intersection;

import "updatable_private_set_intersection/match.proto";

// Client Messages

message PrivateIntersectionClientMessage {
  oneof message_content {
    StartProtocolRequest start_protocol_request = 1;
    ClientRoundOne client_round_one = 2;
    ClientKeyExchange client_key_exchange = 3;
  }

  // For initiating the protocol.
  message StartProtocolRequest {}

  // Message containing the client's set encrypted under the client's keys, and
  // the server's set re-encrypted with the client's key, and shuffled.
  message ClientRoundOne {
    optional bytes public_key = 1;
    optional EncryptedSet encrypted_set = 2;
    optional EncryptedSet reencrypted_set = 3;
  }

  // Client key exchange - Threshold ElGamal & Threshold Paillier
  message ClientKeyExchange {
    optional BigNum elgamal_x = 1;
    optional BigNum paillier_n = 2;
  }

}

// Server Messages.

message PrivateIntersectionServerMessage {
  oneof message_content {
    ServerRoundOne server_round_one = 1;
    ServerKeyExchange server_key_exchange = 2;
  }

  message ServerRoundOne {
    optional EncryptedSet encrypted_set = 1;
  }

  // Server key exchange - Threshold ElGamal & Threshold Paillier
  message ServerKeyExchange {
    optional BigNum elgamal_x = 1;
    optional BigNum paillier_n = 2;    
  }

}
